pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageRedeliveryTracker.java

```diff
@@ -0,0 +1,230 @@
+
+    protected final HashMap<MessageId, Long> ackTimeoutMessages;
+
+    private final RedeliveryBackoff ackTimeoutRedeliveryBackoff;
+
+    public UnAckedMessageRedeliveryTracker(PulsarClientImpl client, ConsumerBase<?> consumerBase,
+                                           ConsumerConfigurationData<?> conf) {
+        super(client, consumerBase, conf);
+        this.ackTimeoutRedeliveryBackoff = conf.getAckTimeoutRedeliveryBackoff();
+        this.ackTimeoutMessages = new HashMap<MessageId, Long>();
+        this.redeliveryMessageIdPartitionMap = new ConcurrentHashMap<>();
+        this.redeliveryTimePartitions = new ArrayDeque<>();
+
+        int blankPartitions = (int) Math.ceil((double) this.ackTimeoutMillis / this.tickDurationInMs);
+        for (int i = 0; i < blankPartitions + 1; i++) {
+            redeliveryTimePartitions.add(new ConcurrentOpenHashSet<>(16, 1));
+        }
+
+        timeout = client.timer().newTimeout(new TimerTask() {
+            @Override
+            public void run(Timeout t) throws Exception {
+                writeLock.lock();
+                try {
+                    ConcurrentOpenHashSet<UnackMessageIdWrapper> headPartition = redeliveryTimePartitions.removeFirst();
+                    if (!headPartition.isEmpty()) {
+                        headPartition.forEach(unackMessageIdWrapper -> {
+                            addAckTimeoutMessages(unackMessageIdWrapper);
+                            redeliveryMessageIdPartitionMap.remove(unackMessageIdWrapper);
+                        });
+                    }
+                    headPartition.clear();
+                    redeliveryTimePartitions.addLast(headPartition);
+                    triggerRedelivery(consumerBase);
+                } finally {
+                    writeLock.unlock();
+                    timeout = client.timer().newTimeout(this, tickDurationInMs, TimeUnit.MILLISECONDS);
+                }
+            }
+        }, this.tickDurationInMs, TimeUnit.MILLISECONDS);
+
+    }
+
+    private void addAckTimeoutMessages(UnackMessageIdWrapper messageIdWrapper) {
+        writeLock.lock();
+        try {
+            MessageId messageId = messageIdWrapper.getMessageId();
+            int redeliveryCount = messageIdWrapper.getRedeliveryCount();
+            long backoffNs = ackTimeoutRedeliveryBackoff.next(redeliveryCount);
+            ackTimeoutMessages.put(messageId, System.currentTimeMillis() + backoffNs);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private void triggerRedelivery(ConsumerBase<?> consumerBase) {
+        if (ackTimeoutMessages.isEmpty()) {
+            return;
+        }
+        Set<MessageId> messageIds = TL_MESSAGE_IDS_SET.get();
+        messageIds.clear();
+
+        try {
+            long now = System.currentTimeMillis();
+            ackTimeoutMessages.forEach((messageId, timestamp) -> {
```

Comment:
Need read lock here?

Reply:
No need to lock，because the caller has a write lock.